process/factory/shard/vmContainerFactory.go

```diff
@@ -60,18 +67,24 @@ func NewVMContainerFactory(args ArgVMContainerFactory) (*vmContainerFactory, err
 	cryptoHook := hooks.NewVMCryptoHook()
 	builtinFunctions := blockChainHookImpl.GetBuiltinFunctionNames()
 
-	return &vmContainerFactory{
-		config:                         args.Config,
-		blockChainHookImpl:             blockChainHookImpl,
-		cryptoHook:                     cryptoHook,
-		blockGasLimit:                  args.BlockGasLimit,
-		gasSchedule:                    args.GasSchedule,
-		builtinFunctions:               builtinFunctions,
-		deployEnableEpoch:              args.DeployEnableEpoch,
-		aheadOfTimeGasUsageEnableEpoch: args.AheadOfTimeGasUsageEnableEpoch,
-		arwenV3EnableEpoch:             args.ArwenV3EnableEpoch,
-		ArwenESDTFunctionsEnableEpoch:  args.ArwenESDTFunctionsEnableEpoch,
-	}, nil
+	factory := &vmContainerFactory{
+		config:                               args.Config,
+		blockChainHookImpl:                   blockChainHookImpl,
+		cryptoHook:                           cryptoHook,
+		blockGasLimit:                        args.BlockGasLimit,
+		gasSchedule:                          args.GasSchedule,
+		builtinFunctions:                     builtinFunctions,
+		epochNotifier:                        args.EpochNotifier,
+		deployEnableEpoch:                    args.DeployEnableEpoch,
+		aheadOfTimeGasUsageEnableEpoch:       args.AheadOfTimeGasUsageEnableEpoch,
+		arwenV3EnableEpoch:                   args.ArwenV3EnableEpoch,
+		ArwenESDTFunctionsEnableEpoch:        args.ArwenESDTFunctionsEnableEpoch,
+		ArwenGasManagementRewriteEnableEpoch: args.ArwenGasManagementRewriteEnableEpoch,
+		container:                            nil,
+	}
+
+	factory.epochNotifier.RegisterNotifyHandler(factory)
```

Comment:
here we have a small problem - RegisterNotifyHandler will call EpochConfirmed where you actually do a replace. factory.container is nil, thus this will panic. maybe in the function EpochConfirmed - you can check whether container was created or not first.